src/blockMonitor/blockCache.ts

```diff
@@ -0,0 +1,230 @@
+        let newBlockStubChain: BlockStubChain;
+        if (parentBlockStubChain === undefined) {
+            newBlockStubChain = BlockStubChain.newRoot({ hash, number, parentHash });
+        } else {
+            newBlockStubChain = parentBlockStubChain.extend({ hash, number, parentHash });
+        }
+        return newBlockStubChain;
+    }
+
+    /**
+     * Returns true it `block` can be added to the cache, that is if either:
+     *   - it is the first block ever seen, or
+     *   - its parent is already in the cache, or
+     *   - it is at depth least `this.maxDepth`.
+     * @param block
+     */
+    public canAddBlock(block: ethers.providers.Block): boolean {
+        return (
+            this.initialHeight === -1 || this.hasBlock(block.parentHash) || block.number <= this.getMinVisibleHeight()
+        );
+    }
+
+    /**
+     * Adds `block`to the cache.
+     * @param block
+     * @returns `true` if the block was added, `false` if the block was not added because too deep or already there.
+     * @throws `ApplicationError` if the block cannot be added because its parent is not in cache.
+     */
+    public addBlock(block: ethers.providers.Block): boolean {
+        if (this.initialHeight === -1) {
+            // First block added, blocks before this point will not be stored.#
+            this.initialHeight = block.number;
+        } else {
+            if (this.blockStubsByHash.has(block.hash)) {
+                // block already in memory
+                return false;
+            }
+
+            if (block.number < this.getMinVisibleHeight()) {
+                // block too deep
+                return false;
+            }
+
+            // If the block's parent is above the minimum visible height, it needs to be added first
+            if (!this.canAddBlock(block)) {
+                throw new ApplicationError("Tried to add a block before its parent block.");
+            }
+        }
+
+        // Update data structures
+
+        // Save block stub
+        const newBlockStub = this.makeBlockStub(block.hash, block.number, block.parentHash);
+        this.blockStubsByHash.set(block.hash, newBlockStub);
+
+        // Add set of transactions for this block hash
+        this.txHashesByBlockHash.set(block.hash, new Set(block.transactions));
+
+        // Index block by its height
+        addItemToKeyedSet(this.blockHashesByHeight, block.number, block.hash);
+
+        // Link all the tx hashes of this block to the block hash
+        for (let txHash of block.transactions || []) {
+            // Add this block among the block containing transaction txHash
+            addItemToKeyedSet(this.blockHashesByTxHash, txHash, block.hash);
+        }
+
+        // If the maximum block height increased, we might have to prune some old info
+        if (this.mMaxHeight < block.number) {
+            this.mMaxHeight = Math.max(this.mMaxHeight, block.number);
```

Comment:
You might add a new block to the cache without increasing the height, e.g. if you're filling the blocks on a different branch that is not the main chain.
Note that the BlockCache does _not_ keep track of the current valid chain, but it uses the height of the tallest block to decide when it can prune blocks. I think this is safe, as usually the currently valid chain should be the longest, although technically it's the maximum-work valid chain one and not the longest one.

Reply:
Yeah, but at this point block.number is guaranteed to be greater than mMaxHeight as you've just checked that this.mMaxHeight < block.number. So Math.max(this.mMaxHeight, block.number) === block.number

Reply:
Right, removed it.